{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf # type: ignore\n",
    "import random\n",
    "def str_vcf_to_genepop(vcf_file, output_file):\n",
    "    \"\"\"Convert VCF format to GenePop format.\"\"\"\n",
    "    vcf_reader = vcf.Reader(filename=vcf_file)\n",
    "    with open(output_file, 'w') as genepop:\n",
    "        # Write title\n",
    "        genepop.write(\"STRvcf2genepop\\n\")\n",
    "        \n",
    "        # Collect locus names and allele count per locus\n",
    "        locus_names = []\n",
    "        unique_ids = set()\n",
    "        locus_allele_count = {}\n",
    "        pos_to_locus_id = {}  # Mapping from record.POS to unique locus_id\n",
    "        for record in vcf_reader:\n",
    "            allele_combinations = {}\n",
    "            ref_repeat_length = len(record.REF)\n",
    "            for sample in record.samples:\n",
    "                gb_data = getattr(sample.data, 'GB', None)\n",
    "                if gb_data:\n",
    "                    allele_diffs = gb_data.split('/')\n",
    "                    # Assuming a pair of allele diffs is available for each sample\n",
    "                    if len(allele_diffs) == 2:\n",
    "                        alleles = []\n",
    "                        for diff in allele_diffs:\n",
    "                            if int(diff) < -100:\n",
    "                                allele = ref_repeat_length  # Handle extreme diffs\n",
    "                            else:\n",
    "                                allele = ref_repeat_length + int(diff)\n",
    "                            alleles.append(f\"{allele:03}\")\n",
    "                        # Combine alleles into a string and count the combination\n",
    "                        allele_key = ''.join(alleles)\n",
    "                        allele_combinations[allele_key] = allele_combinations.get(allele_key, 0) + 1\n",
    "            # Only include loci with more than one allele combination\n",
    "            if len(allele_combinations) > 1:\n",
    "                locus_id = f\"STR_{record.POS}\"\n",
    "                while locus_id in unique_ids:\n",
    "                    # Add a random digit to the end of the locus_id if it's not unique\n",
    "                    locus_id += str(random.randint(0, 9))\n",
    "                unique_ids.add(locus_id)\n",
    "                locus_names.append(locus_id)\n",
    "                locus_allele_count[locus_id] = allele_combinations\n",
    "                pos_to_locus_id[record.POS] = locus_id\n",
    "        genepop.write(\"\\n\".join(locus_names) + \"\\n\")\n",
    "         \n",
    "        # Reset the reader to read samples\n",
    "        vcf_reader = vcf.Reader(filename=vcf_file)\n",
    "        \n",
    "        # Initialize a dictionary to collect genotypes for each sample\n",
    "        samples_genotypes = {}\n",
    "        for record in vcf_reader:\n",
    "            ref_repeat_length = len(record.REF)\n",
    "            # Use the mapping to access the correct locus_id\n",
    "            locus_id = pos_to_locus_id.get(record.POS)\n",
    "            if locus_id and locus_id in locus_allele_count:\n",
    "                for sample in record.samples:\n",
    "                    if sample.sample not in samples_genotypes:\n",
    "                        samples_genotypes[sample.sample] = []\n",
    "                    gb_data = getattr(sample.data, 'GB', None)\n",
    "                    genotype = '000000'  # Default for missing data\n",
    "                    if gb_data:\n",
    "                        allele_diffs = gb_data.split('/')\n",
    "                        alleles = []\n",
    "                        for diff in allele_diffs:\n",
    "                            if int(diff) < -100:\n",
    "                                alleles.append(ref_repeat_length)\n",
    "                            else:\n",
    "                                adjusted_repeat_length = ref_repeat_length + int(diff)\n",
    "                                alleles.append(adjusted_repeat_length)\n",
    "                        genotype = ''.join(f\"{allele:03}\" for allele in alleles)\n",
    "                    samples_genotypes[sample.sample].append(genotype)\n",
    "        \n",
    "        # # Write sample genotypes to file\n",
    "        # for sample, genotypes in samples_genotypes.items():\n",
    "        #     genepop.write(f\"{sample}, \" + \" \".join(genotypes) + \"\\n\")\n",
    "\n",
    "        # Organize samples by prefix and write genotypes with POP labels\n",
    "        prefix_groups = {}\n",
    "        for sample, genotypes in samples_genotypes.items():\n",
    "            prefix = sample.split('_')[0]\n",
    "            if prefix not in prefix_groups:\n",
    "                prefix_groups[prefix] = []\n",
    "            prefix_groups[prefix].append((sample, genotypes))\n",
    "        \n",
    "        for prefix, samples in prefix_groups.items():\n",
    "            genepop.write(\"POP\\n\")\n",
    "            for sample, genotypes in samples:\n",
    "                genepop.write(f\"{sample}, \" + \" \".join(genotypes) + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example usage (commented out):\n",
    "str_vcf_to_genepop(\"SSR.filted.vcf\", \"SSR.filted.gen\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
